KAFKA-7799: Use httpcomponents-client in RestServerTest.

avocader opened the pull request:
The test `org.apache.kafka.connect.runtime.rest.RestServerTest#testCORSEnabled` assumes Jersey client can send restricted HTTP headers(`Origin`).

Jersey client uses `sun.net.www.protocol.http.HttpURLConnection`.
`sun.net.www.protocol.http.HttpURLConnection` drops restricted headers(`Host`, `Keep-Alive`, `Origin`, etc) based on static property `allowRestrictedHeaders`.
This property is initialized in a static block by reading Java system property `sun.net.http.allowRestrictedHeaders`.

So, if classloader loads `HttpURLConnection` before we set `sun.net.http.allowRestrictedHeaders=true`, then all subsequent changes of this system property won't take any effect(which happens if `org.apache.kafka.connect.integration.ExampleConnectIntegrationTest` is executed before `RestServerTest`).
To prevent this, we have to either make sure we set `sun.net.http.allowRestrictedHeaders=true` as early as possible or do not rely on this system property at all.

This PR adds test dependency on `httpcomponents-client` which doesn't depend on `sun.net.http.allowRestrictedHeaders` system property. Thus none of existing tests should interfere with `RestServerTest`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

avocader committed:
KAFKA-7799: Abstract HttpClient to allow sending restricted headers.

avocader committed:
KAFKA-7799: Checkstyle fixes.

avocader committed:
KAFKA-7799: Fixed deserialization of response in executeGet().

wicknicks commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clients/HttpClient.java:
```diff
@@ -0,0 +1,112 @@
+ * All subsequent writes of this property won't take any effect.
+ *
+ */
+public class HttpClient {
```
since this is a Connect specific class, maybe we can call it "ConnectRestClient"? That way we can extend it to contain methods like "createConnector" or "restartTask", all of which hit specific endpoints on the REST API.

avocader replied:
Thanks @wicknicks, I've added `httpcomponents-client` as a test dependency and use it in `RestServerTest`, so we don't have to worry about race condition when setting `sun.net.http.allowRestrictedHeaders` property.

avocader committed:
KAFKA-7799: Updated import-control.

avocader committed:
KAFKA-7799: Fixed allowed imports.

avocader committed:
KAFKA-7799: Added java.net.URISyntaxException to allowed classes.

ijuma commented:
retest this please

ijuma commented:
@gwenshap Can you please review this? This test is failing almost every time (29 out of 30) with Java 11.

avocader committed:
KAFKA-7799: Use httpcomponents-client in RestServerTest.

avocader commented:
Update: in order to not depend on `sun.net.http.allowRestrictedHeaders` system property in unit tests, I've introduced `httpcomponents-client` as a test dependency and updated `RestServerTest` to use it.

avocader committed:
KAFKA-7799: Redundant entries in import control.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestServerTest.java:
```diff
@@ -180,19 +176,26 @@ public void testOptionsDoesNotIncludeWadlOutput() {
-            .accept(MediaType.WILDCARD)
-            .options();
-        Assert.assertEquals(MediaType.TEXT_PLAIN_TYPE, response.getMediaType());
+        HttpRequest request = new HttpOptions("/connectors");
```
Why not make this `HttpOptions request` to avoid the casting later in the method?

avocader replied:
Good catch @rhauch , updated!

rhauch reviewed (commented):
One small comment. It does seem like this is more attractive than dealing with ordering and system properties. One minor nit below.

kkonstantine reviewed (approved):
@avocader 
I like how the updated approach, even though it adds a test dependency, keeps the test code concise and helps cleaning up some boilerplate. 

However, you should update the initial PR message that gives a thorough description of the issue (that will go in the commit message of the PR) in order to reflect the latest approach. 

Again, this seems to be an acceptable workaround in this case, but in a case where you'd need to isolate a similar class that performs static initialization during loading - as we discussed offline - a valid approach would be to create a static factory method and use a minimal custom class loader to return a new instance of this class. 

Otherwise, in terms of implementation LGTM by me.

avocader committed:
KAFKA-7799: Use conrete HttpOptions type to avoid casting.

avocader commented:
Thanks for your review @kkonstantine! I've updated the PR description with the latest approach which I've taken.

rhauch reviewed (approved):
@avocader, thanks for updating the PR description. LGTM.

gwenshap commented:
LGTM

gwenshap commented:
@avocader @rhauch Does this require cherrypicking into another branch?

avocader commented:
@gwenshap this has to be cherry-picked to branches `2.0`, `2.1`, `2.2`. Thanks!

mjsax commented:
@gwenshap Yes, please cherry-pick. Test failed in `2.2` last night again: https://builds.apache.org/blue/organizations/jenkins/kafka-2.2-jdk8/detail/kafka-2.2-jdk8/12/tests

gwenshap commented:
Sorry, missed the responses! Will cherry-pick!

gwenshap commented:
@avocader cherrypicking into 2.1 and 2.0 have some conflicts, if we need this fix on those branches perhaps a new PR would be better (I prefer not to cherrypick when there are conflicts).
